l3kernel/l3pdf.dtx

```diff
@@ -354,12 +411,113 @@
   }
 %    \end{macrocode}
 % \end{macro}
+%
+% \begin{variable}{\c_@@_object_block_int}
+%   Sets the block size used for managing indexed objects.
+%     \begin{macrocode}
+\int_const:Nn \c_@@_object_block_int { 10000 }
+%    \end{macrocode}
+% \end{variable}
+%
+% \begin{macro}{\pdf_object_new_indexed:nn}
+% \begin{macro}{\@@_object_new_indexed:Nn}
+% \begin{macro}{\pdf_object_write_indexed:nnnn, \pdf_object_write_indexed:nnne}
+% \begin{macro}{\@@_object_write_indexed:Nnnn}
+% \begin{macro}[EXP]{\pdf_object_ref_indexed:nn}
+% \begin{macro}[EXP]{\@@_object_ref_indexed:Nn}
+% \begin{macro}[EXP]{\@@_object_indexed_split:nnN}
+%   Using indexed values is much the same, except we have to split up the
+%   value. As that's repetitive, we use a single auxiliary.
+%    \begin{macrocode}
+\cs_new_protected:Npn \pdf_object_new_indexed:nn #1#2
+  {
+    \@@_object_indexed_split:nnN {#1} {#2}
+      \@@_object_new_indexed:Nn
+  }
+\cs_new_protected:Npn \@@_object_new_indexed:Nn #1#2
+  {
+    \@@_backend_object_new:
+    \intarray_if_exist:NF #1
+      {
+        \intarray_new:Nn #1
+          \c_@@_object_block_int
+      }
+    \intarray_gset:Nnn #1 {#2} \g_@@_backend_object_int
+  }
+\cs_new_protected:Npn \pdf_object_write_indexed:nnnn #1#2#3#4
+  {
+    \@@_object_indexed_split:nnN {#1} {#2}
+      \@@_object_new_write_indexed:Nnnn
+        {#3} {#4}
+  }
+\cs_generate_variant:Nn \pdf_object_write_indexed:nnnn { nnne }
+\cs_new_protected:Npn \@@_object_write_indexed:Nnnn #1#2#3#4
+  {
+    \exp_args:Ne \@@_backend_object_write:nnn
+      { \intarray_item:Nn #1 {#2} }
+      {#3} {#4}
+    \bool_gset_true:N \g_@@_init_bool
+  }
+\cs_new:Npn \pdf_object_ref_indexed:nn #1#2
+  {
+    \@@_object_indexed_split:nnN {#1} {#2}
+      \@@_object_ref_indexed:Nn
+  }
+\cs_new:Npn \@@_object_ref_indexed:Nn #1#2
+  {
+    \exp_args:Ne \@@_backend_object_ref:n
+      { \intarray_item:Nn #1 {#2} }
+  }
+\cs_new:Npn \@@_object_indexed_split:nnN #1#2#3
+  {
+    \use:e
+      {
+        \exp_not:N #3
+          \exp_not:c
+            {
+              g_@@_object_ #1 _
+              \int_eval:n
+                { \int_div_truncate:nn {#2} \c_@@_object_block_int }
+              _intarray
+            }
+            {
+              \int_eval:n
+                { \int_mod:nn {#2} \c_@@_object_block_int } + 1
+            }
```

Comment:
This group is the 2nd argument of `\exp_not:N #3`, thus indent it with the same level as `\exp_not:c`.

```suggestion
          {
            \int_eval:n
              { \int_mod:nn {#2} \c_@@_object_block_int } + 1
          }
```